src/engine/renderer/gl_shader.cpp

```diff
@@ -999,10 +1005,54 @@ void GLShaderManager::CompileGPUShaders( GLShader *shader, shaderProgram_t *prog
 				break;
 			}
 
+			// FIXME: add this test: ( strcmp( token, "USE_TCGEN_LIGHTMAP" ) == 0 && r_lightMapping->integer == 0 )
+			// when lightmaps are never used when lightmapping is disabled
+			// see https://github.com/DaemonEngine/Daemon/issues/296 is fixed
+			if ( strcmp( token, "USE_LIGHT_MAPPING" ) == 0 && !r_vertexLighting->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			else if ( strcmp( token, "USE_NORMAL_MAPPING" ) == 0 && r_normalMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			else if ( strcmp( token, "USE_DELUXE_MAPPING" ) == 0 && r_deluxeMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			else if ( strcmp( token, "USE_PHYSICAL_MAPPING" ) == 0 && r_physicalMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			// FIXME: add to the following test: && r_physicalMapping->integer == 0
+			// when reflective specular is implemented for physical mapping too
+			// see https://github.com/DaemonEngine/Daemon/issues/355
+			else if ( strcmp( token, "USE_REFLECTIVE_SPECULAR" ) == 0 && r_specularMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			else if ( strcmp( token, "USE_RELIEF_MAPPING" ) == 0 && r_reliefMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+			else if ( strcmp( token, "USE_HEIGHTMAP_IN_NORMALMAP" ) == 0 && r_reliefMapping->integer == 0 && r_normalMapping->integer == 0 )
+			{
+				program->dummy = true;
+				return;
+			}
+
 			macrosString += Str::Format( "#ifndef %s\n#define %s 1\n#endif\n", token, token );
 		}
 	}
 
+	program->dummy = false;
```

Comment:
What will happen if there was a mistake in identifying which ones will be unused, and someone tries to use it later? I hope there will be an error message explaining what happened.

Reply:
This boolean is there to tell we already know the program will not be used later. The code is not used at all even if we don't keep track of this information.

Reply:
I mean, suppose that due to a bug in calculating which permutations will be needed, you end up trying to activate a shader which was not compiled. There should be a check for this.